readme.md
# Logseq Plugin: *Daily Notice*

- A plugin that displays user messages by day of the week, with the added functionality of reminding about overdue tasks. When the Journal template is loaded, they are displayed.

- Development stage 👷🚧

> This plugin still has room for improvement. Please let me know if you encounter any issues or have any ideas for enhancement.

[![latest release version](https://img.shields.io/github/v/release/YU000jp/logseq-plugin-daily-notice)](https://github.com/YU000jp/logseq-plugin-daily-notice/releases)
[![License](https://img.shields.io/github/license/YU000jp/logseq-plugin-daily-notice?color=blue)](https://github.com/YU000jp/logseq-plugin-daily-notice/LICENSE)
[![Downloads](https://img.shields.io/github/downloads/YU000jp/logseq-plugin-daily-notice/total.svg)](https://github.com/YU000jp/logseq-plugin-daily-notice/releases)
<!-- Published 2023 -->

---

## Feature

---

## Getting Started

### Install from Logseq Marketplace (Coming👷🚧)

- Press [`---`] on the top right toolbar to open [`Plugins`]
- Select marketplace
- Type `Daily` in the search field, select it from the search results and install

### Recommend

- Logseq v0.9.10 or later

### Usage

---

## Showcase / Questions / Ideas / Help

> Go to the [discussion](https://github.com/YU000jp/logseq-plugin-daily-notice/discussions) tab to ask and find this kind of things.

## Author

- GitHub: [YU000jp](https://github.com/YU000jp)

## Prior art & Credit

### Icon

- [icooon-mono.com](https://icooon-mono.com/11304-%e3%82%a2%e3%83%b3%e3%82%b1%e3%83%bc%e3%83%88%e7%94%a8%e7%b4%99%e3%81%ae%e3%82%a2%e3%82%a4%e3%82%b3%e3%83%b3%e7%b4%a0%e6%9d%90/)

---

<a href="https://www.buymeacoffee.com/yu000japan" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-violet.png" alt="🍌Buy Me A Coffee" style="height: 42px;width: 152px" ></a>


CHANGELOG.md
## [1.1.2](https://github.com/YU000jp/logseq-plugin-daily-notice/compare/v1.1.1...v1.1.2) (2023-08-11)


### Bug Fixes

* last release ([d165638](https://github.com/YU000jp/logseq-plugin-daily-notice/commit/d1656381a22d8ed62e076279b81175ed520fcc68))

## [1.1.1](https://github.com/YU000jp/logseq-plugin-daily-notice/compare/v1.1.0...v1.1.1) (2023-07-23)


### Bug Fixes

* **bug:** duplicate popup ([9dec5a2](https://github.com/YU000jp/logseq-plugin-daily-notice/commit/9dec5a2faae43c4b928438829912a31cf6e4b407))

# [1.1.0](https://github.com/YU000jp/logseq-plugin-daily-notice/compare/v1.0.0...v1.1.0) (2023-07-10)


### Features

* **Add:** toolbar & `when Logseq Loaded` to plugin settings ([0fa524b](https://github.com/YU000jp/logseq-plugin-daily-notice/commit/0fa524b81b7175b13dda2b3c12f7d9c9117d8517))

# 1.0.0 (2023-07-10)


### Features

* **Add:** overdue task monitor ([4c44f8f](https://github.com/YU000jp/logseq-plugin-weekday-notice/commit/4c44f8f9d301e14ab04520d4b5f33e8df47564ed))

# 1.0.0 (2023-07-10)


### Features

* **Add:** overdue task monitor ([4c44f8f](https://github.com/YU000jp/logseq-plugin-weekday-notice/commit/4c44f8f9d301e14ab04520d4b5f33e8df47564ed))


src/lib.ts
import { PageEntity } from "@logseq/libs/dist/LSPlugin.user";
import { dailyMessageOpenUI } from "./dailyMessage";
import { setUIoverdue } from "./overdue";

export const getWeekdayString = (targetDay: Date): string => new Intl.DateTimeFormat((logseq.settings?.localizeOrEnglish as string) || "default", { weekday: "long" }).format(targetDay);
export const fromJournals = async (title: string) => {
  if (!logseq.settings) return;
  await logseq.Editor.restoreEditingCursor();
  //Load message box
  if (logseq.settings!.enableMessageBox === true) {
    if (logseq.settings!.toggleSunday === false ||
      logseq.settings!.toggleMonday === false ||
      logseq.settings!.toggleTuesday === false ||
      logseq.settings!.toggleWednesday === false ||
      logseq.settings!.toggleThursday === false ||
      logseq.settings!.toggleFriday === false ||
      logseq.settings!.toggleSaturday === false) return;
    const page = (await logseq.Editor.getPage(title)) as PageEntity | null;
    if (page && page!.journalDay) dailyMessageOpenUI(getDateInputJournalDay(String(page.journalDay)) as Date);
  }
  //Load overdue
  if (logseq.settings!.enableOverdueOnJournalTemplate === true) await setUIoverdue(false);
};//yyyymmdd形式をDateに変換
export const getDateInputJournalDay = (str: string): Date => new Date(
  Number(str.slice(0, 4)),
  Number(str.slice(4, 6)) - 1,
  Number(str.slice(6)),
  0, 0, 0, 0
);

export const closeUI = (key: string) => {
  const element = parent.document.getElementById(logseq.baseInfo.id + `--${key}`) as HTMLDivElement;
  if (element) element.remove();
};
//Dateをyyyymmdd形式の文字列に変換

export const getStringInputDate = (date: Date): string => `${date.getFullYear()}${("0" + (date.getMonth() + 1)).slice(-2)}${("0" + date.getDate()).slice(-2)}`;



src/index.ts
import "@logseq/libs"; //https://plugins-doc.logseq.com/
//import { setup as l10nSetup, t } from "logseq-l10n"; //https://github.com/sethyuan/logseq-l10n
//import ja from "./translations/ja.json";
import { fromJournals } from "./lib";
import { setUIoverdue } from "./overdue";
import { dailyMessageOpenUI } from "./dailyMessage";
import { settingsTemplate } from "./settings";
import { onSettingsChangedForDayOfWeekMessage } from "./dailyMessage";

/* main */
const main = () => {
  // (async () => {
  //   try {
  //     await l10nSetup({ builtinTranslations: { ja } });
  //   } finally {
  /* user settings */
  logseq.useSettingsSchema(settingsTemplate);
  if (!logseq.settings) setTimeout(() => logseq.showSettingsUI(), 300);
  //   }
  // })();

  logseq.App.onTodayJournalCreated(async ({ title }) => await fromJournals(title));

  if (logseq.settings!.enableOverdueLogseqLoaded === true) setTimeout(() => setUIoverdue(false), 300);
  if (logseq.settings!.enableMessageBoxLogseqLoaded === true) setTimeout(() => dailyMessageOpenUI(new Date()), 300);

  
  //for settingUI
  logseq.provideStyle(`
    div#${logseq.baseInfo.id}--overdue span.block-marker {
      cursor: pointer;
      text-decoration: underline;
      padding: 0 0.8em;
      border: 1px solid;
      margin-right: 0.8em;
    }
    div#${logseq.baseInfo.id}--overdue li {
      margin-bottom: 0.5em;
    }
    /* require Logseq v0.9.10 or later */
    body:has(div#root main.ls-right-sidebar-open) div:is(#${logseq.baseInfo.id}--overdue,#${logseq.baseInfo.id}--messageBox) {
      display: none;
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] textarea.form-input {
      height: 12em;
      font-size: unset;
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="backgroundColor"] select option[value="var(--ls-primary-background-color)"] {
      background:var(--ls-primary-background-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="backgroundColor"] select option[value="var(--ls-secondary-background-color)"] {
      background:var(--ls-secondary-background-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="backgroundColor"] select option[value="var(--ls-tertiary-background-color)"] {
      background:var(--ls-tertiary-background-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="backgroundColor"] select option[value="var(--ls-quaternary-background-color)"] {
      background:var(--ls-quaternary-background-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="backgroundColor"] select option[value="var(--ls-table-tr-even-background-color)"] {
      background:var(--ls-table-tr-even-background-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="backgroundColor"] select option[value="var(--ls-block-properties-background-color)"] {
      background:var(--ls-block-properties-background-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="backgroundColor"] select option[value="var(--ls-page-properties-background-color)"] {
      color:var(--ls-page-properties-background-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="fontColor"] select option[value="var(--ls-primary-text-color)"] {
      color:var(--ls-primary-text-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="fontColor"] select option[value="var(--ls-secondary-text-color)"] {
      color:var(--ls-secondary-text-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="fontColor"] select option[value="var(--ls-title-text-color)"] {
      color:var(--ls-title-text-color)
    }
    div#root main div[data-id="${logseq.baseInfo.id}"] div[data-key="fontColor"] select option[value="var(--ls-link-text-color)"] {
      color:var(--ls-link-text-color)
    }
  `);

  //If change settings, show message box
  onSettingsChangedForDayOfWeekMessage();

  // external button on toolbar
  logseq.App.registerUIItem('toolbar', {
    key: 'openOverdue',
    template: `<div><a class="button icon" data-on-click="overdueFromToolbar" title="open overdue task monitor" style="font-size:20px">⏳</a></div>`,
  });
  logseq.App.registerUIItem('toolbar', {
    key: 'openMessageBox',
    template: `<div><a class="button icon" data-on-click="messageBoxFromToolbar" title="open weekday message box" style="font-size:20px">💬</a></div>`,
  });

  logseq.provideModel({
    overdueFromToolbar: () => setUIoverdue(false),
    messageBoxFromToolbar: () => dailyMessageOpenUI(new Date()),
  });


}; /* end_main */



logseq.ready(main).catch(console.error);

src/dailyMessage.ts
import { getWeekdayString } from "./lib";
import { calculateRangeBarForSettingUI } from "./settings";
import { closeUI } from "./lib";
import { LSPluginBaseInfo } from "@logseq/libs/dist/LSPlugin.user";



export const dailyMessageOpenUI = (targetDay: Date, demo?: number) => {
  //曜日を確認
  const dayOfWeek: number = (demo || demo === 0) ? demo : targetDay.getDay();
  //曜日に応じてメッセージを表示
  switch (dayOfWeek) {
    case 0:
      if (logseq.settings!.toggleSunday)
        dailyMessageSetUI(
          getWeekdayString(targetDay),
          logseq.settings!.sunday,
          true,
          false
        );
      break;
    case 1:
      if (logseq.settings!.toggleMonday)
        dailyMessageSetUI(
          getWeekdayString(targetDay),
          logseq.settings!.monday,
          true,
          false
        );
      break;
    case 2:
      if (logseq.settings!.toggleTuesday)
        dailyMessageSetUI(
          getWeekdayString(targetDay),
          logseq.settings!.tuesday,
          true,
          false
        );
      break;
    case 3:
      if (logseq.settings!.toggleWednesday)
        dailyMessageSetUI(
          getWeekdayString(targetDay),
          logseq.settings!.wednesday,
          true,
          false
        );
      break;
    case 4:
      if (logseq.settings!.toggleThursday)
        dailyMessageSetUI(
          getWeekdayString(targetDay),
          logseq.settings!.thursday,
          true,
          false
        );
      break;
    case 5:
      if (logseq.settings!.toggleFriday)
        dailyMessageSetUI(
          getWeekdayString(targetDay),
          logseq.settings!.friday,
          true,
          false
        );
      break;
    case 6:
      if (logseq.settings!.toggleSaturday)
        dailyMessageSetUI(
          getWeekdayString(targetDay),
          logseq.settings!.saturday,
          false,
          false
        );
      break;
  }
};


export const dailyMessageSetUI = (title: string, print: string, timeoutCancel: boolean, demo: boolean) => {
  if (print === undefined) return;
  const width: number = calculateRangeBarForSettingUI(300, 900, logseq.settings?.width as number);
  const height: number = calculateRangeBarForSettingUI(300, 900, logseq.settings?.height as number);
  const backgroundColor = (logseq.settings?.backgroundColor as string) || "var(--ls-primary-background-color)";
  const color = (logseq.settings?.fontColor as string) || "var(--ls-primary-text-color)";
  let left = "unset";
  let bottom = "unset";
  const show = (demo: boolean) => {
    closeUI("messageBox");
    logseq.provideUI({
      key: "messageBox",
      reset: true,
      template: `
    <div style="padding:0.5em">
    ${print}
    </div>
  `,
      style: {
        padding: "0.5em",
        left,
        right: "3px",
        top: "calc(2vh + 50px)",
        bottom,
        width: `${width}px`,
        height: `${height}px`,
        backgroundColor,
        color,
        boxShadow: "1px 2px 5px var(--ls-secondary-background-color)",
        zIndex: demo === true ? "var(--ls-z-index-level-5)" : "9",
      },
      attrs: {
        title: '💬' + title,
      },
    });
    const element = parent.document.getElementById(logseq.baseInfo.id + "--messageBox") as HTMLDivElement;
    if (logseq.settings!.enableMessageBoxTimeout === true) setTimeout(() => {
      let closeCancel: boolean = false;
      element.onclick = () => {
        closeCancel = true;
        if (element) element.style.borderColor = "unset";
      };
      element.onclose = () => closeCancel = true;
      setTimeout(() => {
        if (closeCancel === false && element) element.style.borderColor = "red";
      }, (logseq.settings!.messageBoxTimeout as number) - 2000);
      setTimeout(() => {
        if (closeCancel === false && element) element.remove();
      }, logseq.settings!.messageBoxTimeout as number || 6000);
    }, 100);
    //TODO: なぜか、エラーになってondragendが動作しない
    //https://www.smpl-rfrns.net/DOM/ondragend.html
    // element.addEventListener("dragend", () => {
    //   console.log("moved");
    //     logseq.updateSettings({
    //       UIoverdueWidth: calculateRangeBarForSettingUI(300, 900, Number(element.style.width.slice(0, -2))),
    //       UIoverdueHeight: calculateRangeBarForSettingUI(300, 900, Number(element.style.height.slice(0, -2))),
    //       UIoverdueX: Number(element.style.left.slice(0, -2)),
    //       UIoverdueY: Number(element.style.top.slice(0, -2)),
    //     });
    // });
  };
  timeoutCancel === true ? show(demo) : setTimeout(() => show(demo), 1500);
};


export function onSettingsChangedForDayOfWeekMessage() {
  logseq.onSettingsChanged((newSet: LSPluginBaseInfo["settings"], oldSet: LSPluginBaseInfo["settings"]) => {
    if (newSet.width !== oldSet.width || newSet.height !== oldSet.height) {
      dailyMessageSetUI(
        getWeekdayString(new Date()),
        "Change size of message box",
        true,
        true
      );
    } else if (newSet.backgroundColor !== oldSet.backgroundColor) {
      dailyMessageSetUI(
        getWeekdayString(new Date()),
        "Change background color of message box",
        true,
        true
      );
    } else if (newSet.fontColor !== oldSet.fontColor) {
      dailyMessageSetUI(
        getWeekdayString(new Date()),
        "Change font color of message box",
        true,
        true
      );
    } else if (newSet.localizeOrEnglish !== oldSet.localizeOrEnglish) {
      dailyMessageSetUI(
        getWeekdayString(new Date()),
        "Change language of message box",
        true,
        true
      );
    }
    if (oldSet.toggleMonday === false && newSet.toggleMonday === true) {
      const monday = new Date();
      monday.setDate(monday.getDate() - monday.getDay() + 1);
      dailyMessageSetUI(
        getWeekdayString(monday),
        newSet.monday,
        true,
        true
      );
    } else if (oldSet.toggleTuesday === false && newSet.toggleTuesday === true) {
      const tuesday = new Date();
      tuesday.setDate(tuesday.getDate() - tuesday.getDay() + 2);
      dailyMessageSetUI(
        getWeekdayString(tuesday),
        newSet.tuesday,
        true,
        true
      );
    } else if (oldSet.toggleWednesday === false && newSet.toggleWednesday === true) {
      const wednesday = new Date();
      wednesday.setDate(wednesday.getDate() - wednesday.getDay() + 3);
      dailyMessageSetUI(getWeekdayString(wednesday),
        newSet.wednesday,
        true,
        true
      );
    } else if (oldSet.toggleThursday === false && newSet.toggleThursday === true) {
      const thursday = new Date();
      thursday.setDate(thursday.getDate() - thursday.getDay() + 4);
      dailyMessageSetUI(getWeekdayString(thursday),
        newSet.thursday,
        true,
        true
      );
    } else if (oldSet.toggleFriday === false && newSet.toggleFriday === true) {
      const friday = new Date();
      friday.setDate(friday.getDate() - friday.getDay() + 5);
      dailyMessageSetUI(getWeekdayString(friday),
        newSet.friday,
        true,
        true
      );
    } else if (oldSet.toggleSaturday === false && newSet.toggleSaturday === true) {
      const saturday = new Date();
      saturday.setDate(saturday.getDate() - saturday.getDay() + 6);
      dailyMessageSetUI(getWeekdayString(saturday),
        newSet.saturday,
        true,
        true
      );
    } else if (oldSet.toggleSunday === false && newSet.toggleSunday === true) {
      const sunday = new Date();
      sunday.setDate(sunday.getDate() - sunday.getDay() + 7);
      dailyMessageSetUI(getWeekdayString(sunday),
        newSet.sunday,
        true,
        true
      );
    }
  }
  );
}



src/settings.ts
import { SettingSchemaDesc } from "@logseq/libs/dist/LSPlugin.user";

//for setting UI
export const calculateRangeBarForSettingUI = (min: number, max: number, value: number): number => {
  if (value < 1) value = 1;
  return (value * (max - min) / 100) + min;
};
/* user setting */
// https://logseq.github.io/plugins/types/SettingSchemaDesc.html
export const settingsTemplate: SettingSchemaDesc[] = [
  {
    //messageBox
    key: "enableMessageBox",
    type: "boolean",
    title: "Show message box when today journal is created",
    description: "",
    default: true,
  },
  {
    //messageBox Logseq loaded
    key: "enableMessageBoxLogseqLoaded",
    type: "boolean",
    title: "Show message box when Logseq loaded",
    description: "",
    default: false,
  },
  {
    //width: 500px
    key: "width",
    type: "number",
    title: "Width of message box",
    description: "300 < 600 < 900 [px]",
    default: 30,
    inputAs: "range",
  },
  {
    //height: 500px
    key: "height",
    type: "number",
    title: "Height of message box",
    description: "300 < 600 < 900 [px]",
    default: 20,
    inputAs: "range",
  },
  {
    key: "enableMessageBoxTimeout",
    type: "boolean",
    title: "Enable close message box timeout",
    description: "",
    default: true,
  },
  {
    //close timeout
    key: "messageBoxTimeout",
    type: "enum",
    title: "Close message box timeout [ms]",
    enumChoices: ["8000", "9000", "10000", "12000", "14000", "16000", "18000", "20000"],
    description: "default: 10000",
    default: "10000",
  },
  {
    //background color
    key: "backgroundColor",
    type: "enum",
    title: "Background color (from theme)",
    enumChoices: [
      "var(--ls-primary-background-color)",
      "var(--ls-secondary-background-color)",
      "var(--ls-tertiary-background-color)",
      "var(--ls-quaternary-background-color)",
      "var(--ls-table-tr-even-background-color)",
      "var(--ls-block-properties-background-color)",
      "var(--ls-page-properties-background-color)",
    ],
    description: "default: var(--ls-primary-background-color)",
    default: "var(--ls-primary-background-color)",
  },

  {
    //color
    key: "fontColor",
    type: "enum",
    title: "Font Color (from theme)",
    enumChoices: [
      "var(--ls-primary-text-color)",
      "var(--ls-secondary-text-color)",
      "var(--ls-title-text-color)",
      "var(--ls-link-text-color)",
    ],
    description: "default: var(--ls-primary-text-color)",
    default: "var(--ls-primary-text-color)",
  },
  {
    key: "localizeOrEnglish",
    type: "enum",
    title: "For weekday of the message box title, select localize (your language) or English",
    enumChoices: ["default", "en"],
    description: "",
    default: "default",
  },
  {
    key: "toggleMonday",
    type: "boolean",
    title: "Show Monday Message",
    description: "",
    default: false,
  },
  {
    key: "monday",
    type: "string",
    inputAs: "textarea",
    title: "Message for Monday (Supports HTML instead of markdown)",
    description: "To view the modified content, toggle it off and then on again.",
    default: "",
  },
  {
    key: "toggleTuesday",
    type: "boolean",
    title: "Show Tuesday Message",
    description: "",
    default: false,
  },
  {
    key: "tuesday",
    type: "string",
    inputAs: "textarea",
    title: "Message for Tuesday (Supports HTML instead of markdown)",
    description: "To view the modified content, toggle it off and then on again.",
    default: "",
  },
  {
    key: "toggleWednesday",
    type: "boolean",
    title: "Show Wednesday Message",
    description: "",
    default: false,
  },
  {
    key: "wednesday",
    type: "string",
    inputAs: "textarea",
    title: "Message for Wednesday (Supports HTML instead of markdown)",
    description: "To view the modified content, toggle it off and then on again.",
    default: "",
  },
  {
    key: "toggleThursday",
    type: "boolean",
    title: "Show Thursday Message",
    description: "",
    default: false,
  },
  {
    key: "thursday",
    type: "string",
    inputAs: "textarea",
    title: "Message for Thursday (Supports HTML instead of markdown)",
    description: "To view the modified content, toggle it off and then on again.",
    default: "",
  },
  {
    key: "toggleFriday",
    type: "boolean",
    title: "Show Friday Message",
    description: "",
    default: false,
  },
  {
    key: "friday",
    type: "string",
    inputAs: "textarea",
    title: "Message for Friday (Supports HTML instead of markdown)",
    description: "To view the modified content, toggle it off and then on again.",
    default: "",
  },
  {
    key: "toggleSaturday",
    type: "boolean",
    title: "Show Saturday Message",
    description: "",
    default: false,
  },
  {
    key: "saturday",
    type: "string",
    inputAs: "textarea",
    title: "Message for Saturday (Supports HTML instead of markdown)",
    description: "To view the modified content, toggle it off and then on again.",
    default: "",
  },
  {
    key: "toggleSunday",
    type: "boolean",
    title: "Show Sunday Message",
    description: "",
    default: false,
  },
  {
    key: "sunday",
    type: "string",
    inputAs: "textarea",
    title: "Message for Sunday (Supports HTML instead of markdown)",
    description: "To view the modified content, toggle it off and then on again.",
    default: "",
  },
  {
    key: "enableOverdueOnJournalTemplate",
    type: "boolean",
    title: "Show overdue tasks when journal template is loaded",
    description: "",
    default: true,
  },
  {
    key: "enableOverdueLogseqLoaded",
    type: "boolean",
    title: "Show overdue tasks when Logseq is loaded",
    description: "",
    default: false,
  }
];
